File: python/practice/recursion/mergeSort.py
def merge(arr, low, mid, high):
    i = low
    j = mid+1

    temp = []
    while i <= mid and j <= high:
        if arr[i] < arr[j]:
            temp.append(arr[i])
            i += 1
        else:
            temp.append(arr[j])
            j += 1
        
    while i <= mid:
        temp.append(arr[i])
        i += 1
    
    while j <= high:
        temp.append(arr[j])
        j += 1

    for i in range(len(temp)):
        arr[low + i] = temp[i]


def mergeSort(arr, low, high):
    if low >= high:
         return

    mid = (low + high) // 2
    mergeSort(arr, low, mid)
    mergeSort(arr, mid + 1, high)
    merge(arr, low, mid, high)


arr = [12, 11, 13, 5, 6, 7]

mergeSort(arr, 0, len(arr)-1)

print(arr)

File: CPP/Recursion/mergeSort.cpp
#include<iostream>
#include<vector>
using namespace std;

// divide and merge
void merge(vector<int>& arr, int l, int mid, int r){
    vector<int> temp;
    int i = l, j = mid + 1;
    while(i <= mid && j <= r){
        if(arr[i] < arr[j]){
            temp.push_back(arr[i]);
            i++;
        }else{
            temp.push_back(arr[j]);
            j++;
        }
    }
    while(i <= mid){
        temp.push_back(arr[i]);
        i++;
    }
    while(j <= r){
        temp.push_back(arr[j]);
        j++;
    }
    for(int i = l; i <= r; i++){
        arr[i] = temp[i - l];
    }
}

void mergeSort(vector<int>& arr, int l, int r){
    if(l == r) return;
    int mid = (l + r) / 2;
    mergeSort(arr, l, mid);
    mergeSort(arr, mid + 1, r);
    merge(arr, l, mid, r);
}

int main(){
    vector<int> arr = {9, 1, 2, 7, 6};
    mergeSort(arr, 0, arr.size() - 1);
    cout << "Sorted array: ";
    for(auto it: arr){
        cout << it << " ";
    }
    return 0;
}